server/notifications.go

```diff
@@ -5,91 +5,87 @@
+	mutex *sync.RWMutex
+	cond  *sync.Cond
+
+	// the whole state must be always replaced with `notify`
+	// the struct fields are not thread safe on their own
+	state *state
+
+	// the notifications are stopped when this flag > 0
+	stopped uint32
 }
 
-func (n *notifications) subscribe() (out <-chan change, stop func(), err error) {
-	ch := make(chan change, notificationBufferSize)
-	id, err := uuid.NewV4()
-	if err != nil {
-		return nil, nil, err
-	}
+// getState return the current state or `nil` if the initial state
+// is not set or if the notifications are stopped.
+// This is thread-safe.
+func (n *notifications) getState() *state {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+	return n.state
+}
 
-	n.mutex.Lock()
-	n.subscribers[id] = ch
-	n.mutex.Unlock()
+// wait returns a pointer to the current state or blocks until the
+// first state change or when notifications are stopped.
+func (n *notifications) wait() *state {
+	// before trying to lock we check the flag
+	// it should be less expensive
+	if atomic.LoadUint32(&n.stopped) != 0 {
+		return nil
+	}
 
-	stop = func() {
-		n.mutex.Lock()
-		// we must block till the end of the function so `shutdown`
-		// does not try to close closed channels and panics
-		defer n.mutex.Unlock()
+	n.cond.L.Lock()
+	defer n.cond.L.Unlock()
 
-		// `shutdown` could shut down the notifications before this is executed
-		_, ok := n.subscribers[id]
-		if ok {
-			delete(n.subscribers, id)
-			close(ch)
-			// drain the channel's buffer
-			for range ch {
-			}
-		}
+	// waiting until there is an initial state or everything is stopped
+	for n.state == nil && atomic.LoadUint32(&n.stopped) == 0 {
+		n.cond.Wait()
 	}
 
-	return ch, stop, nil
+	return n.state
 }
 
-// notify notifies all the subscribers.
-// This function should be used asynchronously.
-// If one of subscriber channel's buffer is exhausted it might block
-// until the subscriber reads the previous notification or unsubscribes.
-func (n *notifications) notify(ctx context.Context, value change) {
-	n.mutex.Lock()
-	defer n.mutex.Unlock()
-
-	for _, ch := range n.subscribers {
-		select {
-		case <-ctx.Done():
-			return
-		case ch <- value:
-			continue
-		}
+// notify notifies all the subscribers and returns `true`
+// or does not notify and returns `false` in case there is no update in the state
+func (n *notifications) notify(value state) bool {
+	if atomic.LoadUint32(&n.stopped) != 0 {
+		return false
 	}
+	n.cond.L.Lock()
+	defer n.cond.L.Unlock()
+	if value.Equals(n.state) {
+		return false
+	}
+	n.state = &value
+	n.cond.Broadcast()
+
+	return true
 }
 
-// shutdown closes all the subscriber channels and drains them.
+// shutdown stops waiting for all the subscribers and sets the current state to `nil`.
 func (n *notifications) shutdown() {
-	n.mutex.Lock()
-	defer n.mutex.Unlock()
-
-	for _, ch := range n.subscribers {
-		close(ch)
-		// drain the channel buffer
-		for range ch {
-		}
+	if !atomic.CompareAndSwapUint32(&n.stopped, 0, 1) {
```

Comment:
It seems like you are only using atomics on the shutdown path. Is the extra complexity worth a possible performance gain here? It seems like you could just rely on acquiring `n.cond.L` instead.

Reply:
does it really look so complex? I'd like to point out that any attempt to lock the mutex is potentially blocking and causes competing access. It has to be avoided at all costs. `atomic` is just a CPU instruction for special treatment which will always immediately return. It's not the same thing.

Reply:
I don't think it has to be avoided at all costs, they are used internally all over the Go runtime and nobody notices (example, channels require at least one https://github.com/golang/go/blob/4fc9565ffce91c4299903f7c17a275f0786734a1/src/runtime/chan.go#L28)

I have always found atomics can cause tricky concurrency bugs when you move beyond simple loads and stores. That said this use is realistically fine, resolving.